src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/CollectionBackedStoreProvider.cs

```diff
@@ -0,0 +1,159 @@
+                            // being able to use that functionality requires a lot more code for
+                            // a low-usage scenario.
+                            throw new PlatformNotSupportedException(SR.NotSupported_Export_MultiplePrivateCerts);
+                        }
+
+                        privateCert = cert;
+                    }
+                    else
+                    {
+                        using (SafeX509Handle certHandle = Interop.libcrypto.X509_dup(cert.Handle))
+                        {
+                            if (!Interop.NativeCrypto.PushX509StackField(publicCerts, certHandle))
+                            {
+                                throw Interop.libcrypto.CreateOpenSslCryptographicException();
+                            }
+
+                            // The handle ownership has been transferred into the STACK_OF(X509).
+                            certHandle.SetHandleAsInvalid();
+                        }
+                    }
+                }
+
+                SafeX509Handle privateCertHandle;
+                SafeEvpPkeyHandle privateCertKeyHandle;
+
+                if (privateCert != null)
+                {
+                    OpenSslX509CertificateReader pal = (OpenSslX509CertificateReader)privateCert.Pal;
+                    privateCertHandle = pal.SafeHandle;
+                    privateCertKeyHandle = pal.PrivateKeyHandle ?? SafeEvpPkeyHandle.InvalidHandle;
+                }
+                else
+                {
+                    privateCertHandle = SafeX509Handle.InvalidHandle;
+                    privateCertKeyHandle = SafeEvpPkeyHandle.InvalidHandle;
+                }
+
+                using (SafePkcs12Handle pkcs12 = Interop.libcrypto.PKCS12_create(
+                    password,
+                    null,
+                    privateCertKeyHandle,
+                    privateCertHandle,
+                    publicCerts,
+                    Interop.libcrypto.NID_undef,
+                    Interop.libcrypto.NID_undef,
+                    Interop.libcrypto.PKCS12_DEFAULT_ITER,
+                    Interop.libcrypto.PKCS12_DEFAULT_ITER,
+                    0))
+                {
+                    if (pkcs12.IsInvalid)
+                    {
+                        throw Interop.libcrypto.CreateOpenSslCryptographicException();
+                    }
+
+                    unsafe
+                    {
+                        return Interop.libcrypto.OpenSslI2D(
+                            (handle, b) => Interop.libcrypto.i2d_PKCS12(handle, b),
+                            pkcs12);
+                    }
+                }
+            }
+        }
+
+        public IEnumerable<X509Certificate2> Certificates
+        {
+            get
+            {
+                foreach (X509Certificate2 cert in _certs)
+                {
+                    yield return cert;
```

Comment:
Yeah, I see why you'd want to avoid exposing the collection to callers (even if the callers are internal). Not as part of this PR, but maybe revise the `IStorePal` interface at some point in the future... It looks like all the callers of `Certificates` and `Find` just end up adding each item to a `X509Certificate2Collection`. Perhaps changing `Certificates` and `Find` to methods that take a `X509Certificate2Collection` to add to, and have the `StorePal` handle enumerating an adding to the collection. This would avoid the enumerator allocations (and any intermediate read-only collections).